main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala

```diff
@@ -562,143 +558,136 @@ object Monomorpher {
-      val e = visitExp(exp, env0, subst)
+      // Drop the declaredType and declaredEff.
+      val e = specializeExp(exp, env0, subst)
       MonoAst.Expr.Cast(e, None, None, subst(tpe), subst(eff), loc)
 
     case LoweredAst.Expr.TryCatch(exp, rules, tpe, eff, loc) =>
-      val e = visitExp(exp, env0, subst)
+      val e = specializeExp(exp, env0, subst)
       val rs = rules map {
         case LoweredAst.CatchRule(sym, clazz, body) =>
           // Generate a fresh symbol.
           val freshSym = Symbol.freshVarSym(sym)
           val env1 = env0 + (sym -> freshSym)
-          val b = visitExp(body, env1, subst)
+          val b = specializeExp(body, env1, subst)
           MonoAst.CatchRule(freshSym, clazz, b)
       }
       MonoAst.Expr.TryCatch(e, rs, subst(tpe), subst(eff), loc)
 
     case LoweredAst.Expr.RunWith(exp, effect, rules, tpe, eff, loc) =>
-      val e = visitExp(exp, env0, subst)
+      val e = specializeExp(exp, env0, subst)
       val rs = rules map {
         case LoweredAst.HandlerRule(op, fparams0, body0) =>
           val (fparams, fparamEnv) = specializeFormalParams(fparams0, subst)
           val env1 = env0 ++ fparamEnv
-          val body = visitExp(body0, env1, subst)
+          val body = specializeExp(body0, env1, subst)
           MonoAst.HandlerRule(op, fparams, body)
       }
       MonoAst.Expr.RunWith(e, effect, rs, subst(tpe), subst(eff), loc)
 
     case LoweredAst.Expr.Do(op, exps, tpe, eff, loc) =>
-      val es = exps.map(visitExp(_, env0, subst))
+      val es = exps.map(specializeExp(_, env0, subst))
       MonoAst.Expr.Do(op, es, subst(tpe), subst(eff), loc)
 
     case LoweredAst.Expr.NewObject(name, clazz, tpe, eff, methods0, loc) =>
-      val methods = methods0.map(visitJvmMethod(_, env0, subst))
+      val methods = methods0.map(specializeJvmMethod(_, env0, subst))
       MonoAst.Expr.NewObject(name, clazz, subst(tpe), subst(eff), methods, loc)
 
   }
 
-  /**
-    * Specializes the given pattern `p0` w.r.t. the current substitution.
-    *
-    * Returns the new pattern and a mapping from variable symbols to fresh variable symbols.
-    */
-  private def visitPat(p0: LoweredAst.Pattern, subst: StrictSubstitution)(implicit root: LoweredAst.Root, flix: Flix): (MonoAst.Pattern, Map[Symbol.VarSym, Symbol.VarSym]) = p0 match {
-    case LoweredAst.Pattern.Wild(tpe, loc) => (MonoAst.Pattern.Wild(subst(tpe), loc), Map.empty)
+  /** Specializes `p0` w.r.t. `subst` and returns a renaming of pattern variables. */
+  private def specializePat(p0: LoweredAst.Pattern, subst: StrictSubstitution)(implicit root: LoweredAst.Root, flix: Flix): (MonoAst.Pattern, Map[Symbol.VarSym, Symbol.VarSym]) = p0 match {
+    case LoweredAst.Pattern.Wild(tpe, loc) =>
+      (MonoAst.Pattern.Wild(subst(tpe), loc), Map.empty)
     case LoweredAst.Pattern.Var(sym, tpe, loc) =>
-      // Generate a fresh variable symbol for the pattern-bound variable.
+      // Generate a fresh symbol.
```

Comment:
This loses information.

Reply:
I'd say its redundant, we are in Pattern.Var.